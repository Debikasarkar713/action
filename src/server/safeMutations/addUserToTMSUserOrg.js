import getRethink from 'server/database/rethinkDriver';

const addUserToTMSUserOrg = (userId, teamId, orgId, options = {}) => {
  const {returnChanges, role = null} = options;
  const r = getRethink();
  const now = new Date();

  return r.table('User')
    .get(userId)
    .update((user) => ({
      userOrgs: r.branch(
        user('userOrgs').contains((userOrg) => userOrg('id').eq(orgId)).default(false),
        user('userOrgs'),
        user('userOrgs').append({
          id: orgId,
          role
        })
      ),
      tms: r.branch(
        user('tms').contains(teamId).default(false),
        user('tms'),
        user('tms').default([]).append(teamId)
      ),
      updatedAt: now
    }), {returnChanges});
};

export default addUserToTMSUserOrg;
