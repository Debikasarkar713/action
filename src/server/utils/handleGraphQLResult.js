import jwtDecode from 'jwt-decode';

const setSocketAuth = (socket, authTokenStr) => {
  socket.setAuthToken({
    ...jwtDecode(authTokenStr),
    exp: undefined
  });
};

const handleGraphQLResult = (result, socket) => {
  const {data, errors} = result;
  if (errors) return false;
  if (data.notificationsAdded) {
    const {notifications} = data.notificationsAdded;
    const updateTokenNotification = notifications.find((notification) => notification.authToken);
    if (updateTokenNotification) {
      setSocketAuth(socket, updateTokenNotification.authToken);
      return true;
    }
  } else if (data.acceptTeamInviteNotification) {
    setSocketAuth(socket, data.acceptTeamInviteNotification.authToken);
    return true;
  } else if (data.newAuthToken) {
    // TODO refactor the above 2 clauses so this is the only one
    setSocketAuth(socket, data.newAuthToken);
    return true;
  }
  return false;
};

export default handleGraphQLResult;
