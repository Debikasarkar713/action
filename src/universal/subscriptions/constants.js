// EVENTS
export const EDIT = 'EDIT';
export const LEAVE = 'LEAVE';
export const PRESENT = 'PRESENT';
export const SOUNDOFF = 'SOUNDOFF';

// CHANNELS
export const ARCHIVED_PROJECTS = 'archivedProjects';
export const AGENDA = 'agenda';
export const INVITATIONS = 'invitations';
export const ORG_APPROVALS = 'orgApprovals';
export const ORGANIZATIONS = 'organizations';
export const PRESENCE = 'presence';
export const PROJECTS = 'projects';
export const AGENDA_PROJECTS = 'agendaProjects';
export const TEAM = 'team';
export const TEAM_MEMBERS = 'teamMembers';
