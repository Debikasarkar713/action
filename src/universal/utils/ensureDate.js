import isValidDate from 'universal/utils/isValidDate';

const ensureDate = (maybeDate) => {
  const date = new Date(maybeDate);
  return isValidDate(date) ? date : new Date();
};

export default ensureDate;
