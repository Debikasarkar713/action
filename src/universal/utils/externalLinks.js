// eslint-disable-next-line import/prefer-default-export
export const PRICING_LINK = 'https://www.parabol.co/pricing/';
