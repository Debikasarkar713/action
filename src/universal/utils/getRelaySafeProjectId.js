import fromGlobalId from 'universal/utils/relay/fromGlobalId';

const getRelaySafeProjectId = (projectId) => {
  try {
    const {id: globalId} = fromGlobalId(projectId);
    return globalId;
  } catch (e) {
    return projectId;
  }
};
export default getRelaySafeProjectId;
