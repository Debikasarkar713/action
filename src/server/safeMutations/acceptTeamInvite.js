import adjustUserCount from 'server/billing/helpers/adjustUserCount';
import getRethink from 'server/database/rethinkDriver';
import addUserToTMSUserOrg from 'server/safeMutations/addUserToTMSUserOrg';
import insertNewTeamMember from 'server/safeMutations/insertNewTeamMember';
import getTeamInviteNotifications from 'server/safeQueries/getTeamInviteNotifications';
import {auth0ManagementClient} from 'server/utils/auth0Helpers';
import getPubSub from 'server/utils/getPubSub';
import {ADD_USER} from 'server/utils/serverConstants';
import {
  ADD_TO_TEAM, JOIN_TEAM, NEW_AUTH_TOKEN, NOTIFICATIONS_ADDED,
  NOTIFICATIONS_CLEARED
} from 'universal/utils/constants';
import {getUserId} from 'server/utils/authorization';
import tmsSignToken from 'server/utils/tmsSignToken';

const acceptTeamInvite = async (teamId, authToken, email) => {
  const r = getRethink();
  const now = new Date();
  const userId = getUserId(authToken);
  const {team: {orgId, name: teamName}, user} = await r({
    team: r.table('Team').get(teamId).pluck('orgId', 'name'),
    user: r.table('User').get(userId)
  });
  const userOrgs = user.userOrgs || [];
  const userTeams = user.tms || [];
  const userInOrg = Boolean(userOrgs.find((org) => org.id === orgId));
  const tms = [...userTeams, teamId];
  const {expireInviteNotificationIds} = await r({
    // add the team to the user doc
    userUpdate: addUserToTMSUserOrg(userId, teamId, orgId),
    newTeamMember: insertNewTeamMember(userId, teamId),
    // find all possible emails linked to this person and mark them as accepted
    expireEmailInvitations: r.table('Invitation')
      .getAll(email, {index: 'email'})
      .update({
        acceptedAt: now,
        // flag the token as expired so they cannot reuse the token
        tokenExpiration: new Date(0),
        updatedAt: now
      }),
    expireInviteNotificationIds: getTeamInviteNotifications(orgId, teamId, [email])
      .delete({returnChanges: true})('changes')
      .map((change) => change('old_val')('id'))
      .default([])
  });

  if (!userInOrg) {
    await adjustUserCount(userId, orgId, ADD_USER);
  }
  auth0ManagementClient.users.updateAppMetadata({id: userId}, {tms});

  // Clear all invitiation notifications
  if (expireInviteNotificationIds.length > 0) {
    const notificationsCleared = {deletedIds: expireInviteNotificationIds};
    getPubSub().publish(`${NOTIFICATIONS_CLEARED}.${userId}`, {notificationsCleared});
  }

  // Tell the team who just joined
  const notificationsAdded = {
    notifications: [{
      type: JOIN_TEAM,
      teamName,
      preferredName: user.preferredName || user.email
    }]
  };
  getPubSub().publish(`${NOTIFICATIONS_ADDED}.${teamId}`, {notificationsAdded});

  // Send the new team member a welcome & a new token
  const newAuthToken = tmsSignToken(authToken, tms);
  const addedToTeam = {
    // no ID to rule out permanent notification. they already know, they are the ones who did it!
    authToken: newAuthToken,
    orgId,
    startAt: now,
    type: ADD_TO_TEAM,
    teamName,
    teamId
  };

  getPubSub().publish(`${NEW_AUTH_TOKEN}.${userId}`, {newAuthToken});
  getPubSub().publish(`${NOTIFICATIONS_ADDED}.${userId}`, {notificationsAdded: {notifications: [addedToTeam]}});
  return addedToTeam;
};

export default acceptTeamInvite;
