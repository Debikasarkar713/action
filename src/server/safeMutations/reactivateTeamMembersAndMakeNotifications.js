import getRethink from 'server/database/rethinkDriver';
import makeReactivationNotifications from 'server/safeMutations/helpers/makeReactivationNotifications';
import shortid from 'shortid';
import {ADD_TO_TEAM, NEW_AUTH_TOKEN} from 'universal/utils/constants';
import getPubSub from 'server/utils/getPubSub';
import tmsSignToken from 'server/utils/tmsSignToken';

const reactivateTeamMembersAndMakeNotifications = async (invitees, inviter, teamMembers) => {
  if (invitees.length === 0) return [];
  const {orgId, teamId, teamName} = inviter;
  const r = getRethink();
  const now = new Date();
  const userIds = invitees.map(({userId}) => userId);
  const teamMemberIds = userIds.map((userId) => `${userId}::${teamId}`);
  const {reactivatedUsers} = await r({
    reactivatedTeamMembers: r.table('TeamMember')
      .getAll(r.args(teamMemberIds), {index: 'id'})
      .update({isNotRemoved: true}),
    reactivatedUsers: r.table('User')
      .getAll(r.args(userIds))
      .update((user) => {
        return user.merge({
          tms: user('tms').append(teamId)
        });
      }, {returnChanges: true})('changes')('new_val')
  });
  const notifications = reactivatedUsers.map((user) => ({
    id: shortid.generate(),
    type: ADD_TO_TEAM,
    startAt: now,
    orgId,
    userIds: [user.id],
    teamId,
    teamName
  }));
  await r.table('Notification').insert(notifications);
  reactivatedUsers.forEach(({id: userId, tms}) => {
    getPubSub().publish(`${NEW_AUTH_TOKEN}.${userId}`, {newAuthToken: tmsSignToken({sub: userId}, tms)});
  });
  return makeReactivationNotifications(notifications, reactivatedUsers, teamMembers, inviter);
};

export default reactivateTeamMembersAndMakeNotifications;
