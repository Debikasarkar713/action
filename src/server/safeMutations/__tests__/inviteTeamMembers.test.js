import DynamicSerializer from 'dynamic-serializer';
import MockDate from 'mockdate';
import MockDB from 'server/__tests__/setup/MockDB';
import {__now} from 'server/__tests__/setup/mockTimes';
import newInvitee from 'server/__tests__/utils/newInvitee';
import * as publishNotifications from 'server/utils/publishNotifications';
import * as createPendingApprovals from 'server/safeMutations/createPendingApprovals';
import inviteTeamMembers from 'server/safeMutations/inviteTeamMembers';
import * as reactivateTeamMembersAndMakeNotifications from 'server/safeMutations/reactivateTeamMembersAndMakeNotifications';
import * as removeOrgApprovalAndNotification from 'server/safeMutations/removeOrgApprovalAndNotification';
import * as sendTeamInvitations from 'server/safeMutations/sendTeamInvitations';
import {REACTIVATED} from 'universal/utils/constants';

MockDate.set(__now);
console.error = jest.fn();

describe('inviteTeamMembers', () => {
  test('reactivates a team member that has been invited back to the team', async () => {
    // SETUP
    sendTeamInvitations.default = jest.fn(() => []);
    reactivateTeamMembersAndMakeNotifications.default = jest.fn(() => []);
    removeOrgApprovalAndNotification.default = jest.fn(() => []);
    createPendingApprovals.default = jest.fn(() => []);
    publishNotifications.default = jest.fn(() => []);

    const mockDB = new MockDB();
    const invitee = newInvitee();
    await mockDB.init()
      .newUser({name: 'Leader of One', email: invitee.email})
      .newTeamMember()
      .teamMember(8, {isNotRemoved: false});
    const firstUser = mockDB.db.user[0];
    const invitees = [{email: invitee.email}];
    const teamId = mockDB.context.team.id;
    // TEST
    const {results} = await inviteTeamMembers(invitees, teamId, firstUser.id);

    // VERIFY
    const res = results[0];
    expect(reactivateTeamMembersAndMakeNotifications.default).toHaveBeenCalledTimes(1);
    expect(res.preferredName).toEqual('Leader of One');
    expect(res.email).toEqual(invitee.email);
    expect(res.result).toEqual(REACTIVATED);
  });

  test('invites a new person via notification and email', async () => {
    // SETUP
    sendTeamInvitations.default = jest.fn(() => []);
    reactivateTeamMembersAndMakeNotifications.default = jest.fn(() => []);
    removeOrgApprovalAndNotification.default = jest.fn(() => []);
    createPendingApprovals.default = jest.fn(() => []);
    publishNotifications.default = jest.fn(() => []);
    const dynamicSerializer = new DynamicSerializer();
    const mockDB = new MockDB();
    const invitee = newInvitee();
    await mockDB.init()
      .newOrg({name: 'Sad Sacks, Inc.'})
      .newTeam({name: 'The Lonely Ones'})
      .newUser({name: 'Leader of One', email: invitee.email})
      .newTeamMember({isLead: true});
    const firstUser = mockDB.db.user[0];
    const invitees = [{email: mockDB.context.user.email}];
    const teamId = mockDB.context.team.id;
    // TEST
    const res = await inviteTeamMembers(invitees, teamId, firstUser.id);

    // VERIFY
    expect(sendTeamInvitations.default).toHaveBeenCalledTimes(1);
    expect(removeOrgApprovalAndNotification.default).toHaveBeenCalledTimes(1);
    const staticRes = dynamicSerializer.toStatic(res, ['results.email']);
    expect(staticRes).toMatchSnapshot();
  });

  test('asks the billing leader for approval', async () => {
    // SETUP
    sendTeamInvitations.default = jest.fn(() => []);
    reactivateTeamMembersAndMakeNotifications.default = jest.fn(() => []);
    removeOrgApprovalAndNotification.default = jest.fn(() => []);
    createPendingApprovals.default = jest.fn(() => []);
    publishNotifications.default = jest.fn(() => []);
    const dynamicSerializer = new DynamicSerializer();
    const mockDB = new MockDB();
    const invitee = newInvitee();
    await mockDB.init();
    const secondUser = mockDB.db.user[1];
    const invitees = [{email: invitee.email}];
    const teamId = mockDB.context.team.id;

    // TEST
    const res = await inviteTeamMembers(invitees, teamId, secondUser.id);

    // VERIFY
    expect(createPendingApprovals.default).toHaveBeenCalledTimes(1);
    const staticRes = dynamicSerializer.toStatic(res, ['results.email']);
    expect(staticRes).toMatchSnapshot();
  });
});
