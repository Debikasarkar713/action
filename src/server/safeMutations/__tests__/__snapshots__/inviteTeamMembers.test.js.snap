// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`inviteTeamMembers asks the billing leader for approval 1`] = `
Object {
  "results": Array [
    Object {
      "email": 0,
      "preferredName": undefined,
      "result": "PENDING_APPROVAL",
    },
  ],
}
`;

exports[`inviteTeamMembers invites a new person via notification and email 1`] = `
Object {
  "results": Array [
    Object {
      "email": 0,
      "preferredName": "Leader of One",
      "result": undefined,
    },
  ],
}
`;
